### PR TITLE
Classifier: refactor polygon tool

### PR DESCRIPTION
**Pull request by eatyourgreens**
Some small refactors to the polygon drawing tool:
- a `FixedNumber` type which represents numbers that are fixed to 2 decimal places.
- an `initialDrag` method which copies the initial drag for line tools.
- refactor the polygon path to use a mapped array of points.
- refactor `polygon.setCoordinates` to match other drawing tools.
- refactor Polygon drag handles to use `point.moveTo({ x, y })` for better performance.
- refactor the first point to use `DragHandle`. The starting point can now be moved!
- double click to finish the polygon from a point.
- add Polygon stories.

This PR also fixes a bug where creating an invalid polygon can crash the classifier, when the invalid mark is automatically deleted but the code tries to call `mark.finish()` on the deleted mark.

## Package
lib-classifier

## How to Review
The polygon tool story, in the classifier storybook, might be the easiest way to review this.

Alternatively, I have a test workflow for the drawing tools, which includes the polygon tool.
https://localhost:8080/?project=908&workflow=3370

https://user-images.githubusercontent.com/59547/177343336-59f8bf5b-64f5-4582-a9c7-fb94d0ca7c03.mov

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## Refactoring
- [ ] The PR creator has described the reason for refactoring
- [ ] The refactored component(s) continue to work as expected

### CODE REVIEW

**Comment by eatyourgreens** [comment]
The polygon tool here works differently from the original polygon tool (https://master.pfe-preview.zooniverse.org/projects/eatyourgreens/-project-testing-ground/classify). @shaunanoordin is that intentional, or should I update this polygon tool to match the original?

**Comment by eatyourgreens** [comment]
I've made some changes here, to bring the tool into line with the PFE version:
- the first point can now be dragged into a new position.
- double click while placing a point to place the point and also finish the polygon.

Those can be reverted, if the new polygon tool is supposed to work differently from the PFE version.

**Comment by eatyourgreens** [comment]
I'm adding this to the Freehand Line project because the `FixedNumber` type, introduced here, is also used by freehand lines.

**Comment by shaunanoordin** [comment]
> Notes on 2 decimal point precision:
> On PFE, Polygon annotations have unlimited precision, so we would see coordinates like { x: 12.3456789, y: 12.3456789 }, which caused extremely bloated (data size-wise) Classifications.
> The move to using 2 decimal points precision in FEM is intentional 👍 Project owners don't need too much precision, and we need reasonably-sized Classifications.

❗ Wait, you know what? I just realised something, and I might need to contradict myself from 10 minutes ago. Why does Polygon need to have 2 d.p. precision? Wasn't it the **freehand drawing tools** that were primarily causing bloat? ❓

**Comment by eatyourgreens** [comment]
TBH, 8 digit precision is probably overkill for all the drawing tools. I added the `FixedNumber` type here so that it’s available to any drawing tool.

File: packages/lib-classifier/src/plugins/drawingTools/models/marks/Polygon/Polygon.js [hunk 1]
```diff
@@ -1,19 +1,27 @@
 import { getParentOfType, types } from 'mobx-state-tree'
 import { Polygon as PolygonComponent } from '@plugins/drawingTools/components/'
+import { FixedNumber } from '@plugins/drawingTools/types/'
 import { Mark } from '@plugins/drawingTools/models/marks'
 import { PolygonTool } from '@plugins/drawingTools/models/tools'
-import roundCoordinates from '@helpers/roundCoordinates'
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/drawingTools/models/marks/Polygon/Polygon.js:5]: I’m not sure. #3393 uses `FixedNumber` for freehand lines.

**Pull request merged**